plugins/grid-bookmark/src/index.ts

```diff
@@ -0,0 +1,148 @@
+        if (pluggableElement.name === 'LinearGenomeView') {
+          const { stateModel } = pluggableElement as ViewType
+          const newStateModel = stateModel.extend(
+            (self: LinearGenomeViewModel) => {
+              const superMenuItems = self.menuItems
+              const superRubberBandMenuItems = self.rubberBandMenuItems
+              return {
+                actions: {
+                  activateBookmarkWidget() {
+                    const session = getSession(self)
+                    if (isSessionModelWithWidgets(session)) {
+                      let bookmarkWidget = session.widgets.get('GridBookmark')
+                      if (!bookmarkWidget) {
+                        bookmarkWidget = session.addWidget(
+                          'GridBookmarkWidget',
+                          'GridBookmark',
+                          { view: self },
+                        )
+                      }
+
+                      session.showWidget(bookmarkWidget)
+                      return bookmarkWidget
+                    }
+
+                    throw new Error('Could not open bookmark widget')
+                  },
+
+                  bookmarkCurrentRegion() {
+                    const selectedRegions = self.getSelectedRegions(
+                      self.leftOffset,
+                      self.rightOffset,
+                    )
+                    const firstRegion = selectedRegions[0]
+                    const session = getSession(self)
+                    if (isSessionModelWithWidgets(session)) {
+                      const { widgets } = session
+                      let bookmarkWidget = widgets.get('GridBookmark')
+                      if (!bookmarkWidget) {
+                        this.activateBookmarkWidget()
+                        bookmarkWidget = widgets.get('GridBookmark')
+                      }
+                      // @ts-ignore
+                      bookmarkWidget.addBookmark(firstRegion)
+                    }
+                  },
+                },
+                views: {
+                  menuItems() {
+                    const menuItems = superMenuItems()
+                    menuItems.push(
+                      { type: 'divider' },
+                      {
+                        label: 'Open bookmark widget',
+                        icon: BookmarksIcon,
+                        // @ts-ignore
+                        onClick: self.activateBookmarkWidget,
+                      },
+                      {
+                        label: 'Bookmark current region',
+                        icon: BookmarkIcon,
+                        // @ts-ignore
+                        onClick: self.bookmarkCurrentRegion,
+                      },
+                    )
+                    return menuItems
+                  },
+
+                  rubberBandMenuItems() {
+                    const rubberBandMenuItems = superRubberBandMenuItems()
```

Comment:
can you use [...superMenuItems, rest of the stuff]? seems like a more immutable way to handle system

Reply:
(push would mutate the array to be specific)

Reply:
Switched over to spreading the values and it seems to work. Just curious, what is the value proposition in it being immutable here?

Reply:
in this case, it would probably be fine because we know what superRubberBandMenuItems is (it is a function we made that returns a plain array that is probably safe to mutate directly) but it is not the best assumption to make in general. a case where this bit us recently was that SimpleFeature modified the data that was inputted to it in a way that caused it to fail (#2056)